src/clipboard_handlers/merge_clipboard.ts

```diff
@@ -1,95 +1,89 @@
-import { getPasteZones } from "../helpers/clipboard/clipboard_helpers";
+import { MinimalClipboardData } from "../plugins/ui_stateful";
 import {
   CellPosition,
-  ClipboardCell,
+  ClipboardCellData,
   ClipboardOptions,
   ClipboardPasteTarget,
-  CommandResult,
   HeaderIndex,
+  Maybe,
+  Merge,
   UID,
-  Zone,
 } from "../types";
 import { AbstractCellClipboardHandler } from "./abstract_cell_clipboard_handler";
 
-interface ClipboardContent {
-  cells: ClipboardCell[][];
-  zones: Zone[];
-  sheetId: UID;
+interface ClipboardContent extends MinimalClipboardData {
+  merges: Maybe<Merge>[][];
 }
 
 export class MergeClipboardHandler extends AbstractCellClipboardHandler<
   ClipboardContent,
-  ClipboardCell
+  Maybe<Merge>
 > {
-  isPasteAllowed(sheetId: UID, target: Zone[], content: ClipboardContent): CommandResult {
-    if (!("cells" in content)) {
-      return CommandResult.Success;
+  copy(data: ClipboardCellData): ClipboardContent | undefined {
+    if (!data.zones.length) {
+      return;
     }
-    const clipboardHeight = content.cells.length;
-    const clipboardWidth = content.cells[0].length;
-    for (const zone of getPasteZones(target, content.cells)) {
-      if (this.getters.doesIntersectMerge(sheetId, zone)) {
-        if (
-          target.length > 1 ||
-          !this.getters.isSingleCellOrMerge(sheetId, target[0]) ||
-          clipboardHeight * clipboardWidth !== 1
-        ) {
-          return CommandResult.WillRemoveExistingMerge;
-        }
+
+    const sheetId = this.getters.getActiveSheetId();
+    const { rowsIndexes, columnsIndexes } = data;
+    const merges: Maybe<Merge>[][] = [];
+
+    for (const row of rowsIndexes) {
+      const mergesInRow: Maybe<Merge>[] = [];
+      for (const col of columnsIndexes) {
+        const position = { col, row, sheetId };
+        mergesInRow.push(this.getters.getMerge(position));
       }
+      merges.push(mergesInRow);
     }
-    return CommandResult.Success;
+    return { merges };
   }
 
   /**
    * Paste the clipboard content in the given target
    */
   paste(target: ClipboardPasteTarget, content: ClipboardContent, options: ClipboardOptions) {
-    if (options?.isCutOperation || !("zones" in target) || !target.zones.length) {
+    if (
+      !content.merges ||
```

Comment:
AFAICT, `content.merges` is always truthy. Remember that an empty array `[]` is truthy in javascript :wink: 

In this case, I would not add any check here. If the array is empty, it won't do anything anyway.

Reply:
The paste function here handles the pasting of merges. It doesn't make sense to move forward if they are no merges 🙂

Reply:
> It doesn't make sense to move forward if they are no merges

but checking adds (useless) code. You don't add a check before every for loop to check if the iterated array is empty, right ? That's the same here. You can just let the flow go and it's perfectly fine.

No check is less code. If there's less code, it's easier to read and less bug prone :wink: